sympy/core/function.py

```diff
@@ -441,6 +441,8 @@ def _diff_wrt(self):
     @cacheit
     def __new__(cls, *args, **options):
         # Handle calls like Function('f')
+        if (hasattr(args[0],'applyfunc')): # simple override for Equation class.
+            return(args[0].applyfunc(cls,*args[1:],**options))
```

Comment:
I'm not sure about this. It's a good way of implementing it but conceptually it irks me slightly that non-Expr args are being sanctioned in the args of objects that should only have Expr args...

Maybe (this needs more thought and should be discussed in its own issue) there should in general be an `_eval_func` method which can be used to override the operation of defined functions.

Reply:
I agree. This was the way I could find to do this that minimized the impact on code external to equation.py. I am also very unhappy about messing with the args list. But could not think of another other way of avoiding an inappropriate reference to the equation object. I suppose the `.applyfunc` or whatever replaces it could take one dummy argument that it ignores.

Reply:
I have modified this to make sure it is only used on objects of class `Equation`. Unfortunately, I have not come up with a cleaner way of adjusting the arguments to pass to `.applyfunc`.

Reply:
This is related to issue #20695. To allow both `Eqn.method()` and `method(Eqn)` to work, I think we have to accept `Eqn` anywhere we accept  `Expr`. It will make little sense to most users if they can only use both forms on `Expr`. Additionally, I find that beginners are much more comfortable with the `method(Eqn)` syntax.

How we go about making this work over the long term can be changed. However, writing a different kind of override into every function would severely delay the inclusion of the `Eqn` class in sympy. I am also not sure about the practicality of having specific overrides for each function over the long term. As sympy grows the coding burden associated with any new extension could be significant. 

I ask that we approve of this general override for now.

Reply:
I think it is better just to limit the features of `Eqn` for now and focus on getting the basic functionality included in sympy. I'm not sure that we want to allow `cos(eqn)` at all. There are other ways to do this like `eqn.apply(cos)` so it's not essential. Once `Eqn` is in and users etc can test it we can explore what else can be done with it.

The main thing that this class solves is awkward constructs like
```
eq = Eq(eq.lhs.collect(eq.atoms(cos)), eq.rhs.collect(eq.atoms(cos))
eq = Eq(*(side.collect(eq.atoms(cos)) for side in [eq.lhs, eq.rhs]))
eq = Eq(eq.lhs, factor(eq.rhs))
```
Compared to that I think that `eqn.apply(cos)` is already a major improvement.

Reply:
I can take it out, but I think it is a mistake for two reasons:
1. It reduces the ease of use. By this I mean it increases the cognitive load of using the tool. With increased cognitive load it becomes less useful because rather than thinking about the math sympy is being used to do the user has to use limited short-term memory space to think about specialized syntax. This is easier for people with more python/programming experience as much of the syntax processing is off-loaded to long-term memorized patterns. Reducing the ease of use reduces the likelihood that people will find sympy useful. I think the goal is to have sympy be useful to as many people as possible.
2. From the perspective of a teacher, I think we are at a critical juncture. Sympy and other symbolic math tools have reached the stage where they have the potential to be a useful tool to help people understand more sophisticated mathematics by allowing them to let the computer do the algebra. I have many students that are capable of dealing with the complexities of the quantum mechanics and statistical thermodynamics I teach, but get bogged down by silly or sloppy algebra mistakes. If the cognitive load is too high then they cannot take advantage of the tool to reduce the impact of those mistakes. The goal is to use this just as people use a calculator or computer to reduce the impact arithmetic errors.

This mechanism makes the behavior of sympy's mathematical functions consistent with the way python methods behave when they operate on an `Equation` or `Expr`. I had to be careful to make `.apply(method)`  work correctly with arbitrary python functions, but `method(Equation)` works by default because of the underlying python and sympy structure. I think it is a bad idea to have inconsistent behavior. It confuses the user. As this only impacts behavior of `Equation` and can be modified or removed without impacting any other part of sympy, I personally think it should be included. We can adjust it if people find it causes problems. As I have been using this with a class for the last semester, I suspect it will not cause problems. However, a larger test community would be better.

I do agree that `eqn.apply(cos)` is much better than the code necessary to make things like this work with `Eq`. But `cos(eqn)` is even better and much easier for people who are not familiar with the concept of object oriented programming. I think the goal is for sympy to be a generally useful tool, not just a tool for coding adept members of the STEM community.

Reply:
I don't necessarily disagree but I think it needs careful consideration and it needs to be agreed by others (not just me). Right now I would really like to get the bulk of this in and merged so that people can get used to it and we can consider future extensions separately, each on their own merit.

This PR is already over 100 posts and is 600 new lines in the codebase. I think if we can keep it to the uncontroversial parts for now then it will put us in a better position to consider details like this afterwards.

Reply:
If this is going to turn into a blocker for acceptance, I can remove it. Then I will make a separate pull request to add this behavior. 

However, this behavior and the application of binary operators (+, - etc) symmetrically on both sides was the key behavior I was trying to get when I originally built my wrapper for sympy. The majority of additional work on the code has been for coding style, testing, efficiency, accommodating requests for alternate syntax from others and to properly interact with `sympy.printing`.

Removal will leave me still having to provide a separate extension to sympy to get the behavior I want and that more naive users (my students and colleagues) expect. It is not a big deal for me to provide a custom version of sympy on systems I control (e.g. my class jupyterhub). However, my students and colleagues will find it much more difficult to set things up the same way, even if I provide a wrapper that can be installed via `pip`.

Who else should be brought into this conversation, irrespective of whether it is completed here or in a separate pull request?